fn/async.go

```diff
@@ -0,0 +1,110 @@
+package fn
+
+import (
+	"context"
+	"errors"
+)
+
+// Async represents an action whose results are being computed outside this
+// control thread.
+type Async[A any] interface {
+	// Await may be called to block on the completion of the remote
+	// computation.
+	Await() (*A, error)
+
+	// Cancel may be called to abandon the remote computation.
+	Cancel()
+}
+
+var (
+	// ContextCancelled indicates that this Async was cancelled because the
+	// surrounding context was being torn down.
+	ContextCancelled = errors.New("context cancelled")
+)
+
+// RunAsyncWithCancel is a function that will transform a synchronous function
+// and make it a cancellable asynchronous call. This can be useful if the
+// function in the argument can block indefinitely and the calling context needs
+// a way to abandon waiting for it to complete.
+func RunAsyncWithCancel[A any](ctx context.Context,
+	f func(ctx context.Context) (*A, error)) Async[A] {
```

Comment:
why use a pointer as a result ?

Reply:
Great question! I actually would prefer not to! I made this choice because it conforms to many go idioms we already have. We could do just `(A, error)` and provide the zero value. This of course still allows pointers as you can always let A = *B. I think we can make that change.

Reply:
Yes I prefer to keep the use of pointer minimal, learnt from you 😅